app/components/navigation_header/index.tsx

```diff
@@ -58,58 +61,73 @@ const NavigationHeader = ({
         return {
             height: Math.max(height, minHeight),
             minHeight,
             maxHeight: largeHeight + insets.top + MAX_OVERSCROLL,
         };
     });
 
+    const minScrollValue = useDerivedValue(() => scrollValue?.value || 0, [scrollValue]);
+    const heightOffset = useDerivedValue(() => largeHeight - defaultHeight, [largeHeight, defaultHeight]);
+
+    const translateY = useDerivedValue(() => (
+        lockValue?.value ? -lockValue.value : Math.min(-minScrollValue.value, heightOffset.value)
+    ), [lockValue?.value, minScrollValue.value]);
+
+    const searchTopMargin = useAnimatedStyle(() => {
+        const min = heightOffset.value;
+        const margin = Math.min(-Math.min(minScrollValue.value, min), min);
+        const unlockedBottomMargin = margin + ViewConstants.UNLOCKED_SEARCH_BOTTOM_MARGIN;
+        const bottomMargin = lockValue?.value ? -lockValue?.value : unlockedBottomMargin;
+        return {
+            marginTop: bottomMargin - ViewConstants.SEARCH_INPUT_HEIGHT,
+        };
+    }, [lockValue?.value, heightOffset.value, minScrollValue.value]);
+
     return (
         <>
             <Animated.View style={[styles.container, containerHeight]}>
                 <Header
                     defaultHeight={defaultHeight}
                     hasSearch={hasSearch}
                     isLargeTitle={isLargeTitle}
-                    largeHeight={largeHeight}
+                    heightOffset={heightOffset}
                     leftComponent={leftComponent}
                     onBackPress={onBackPress}
                     onTitlePress={onTitlePress}
                     rightButtons={rightButtons}
+                    lockValue={lockValue}
                     scrollValue={scrollValue}
                     showBackButton={showBackButton}
                     subtitle={subtitle}
                     subtitleCompanion={subtitleCompanion}
                     theme={theme}
                     title={title}
-                    top={insets.top}
                 />
                 {isLargeTitle &&
-                <NavigationHeaderLargeTitle
-                    defaultHeight={defaultHeight}
-                    hasSearch={hasSearch}
-                    largeHeight={largeHeight}
-                    scrollValue={scrollValue}
-                    subtitle={subtitle}
-                    theme={theme}
-                    title={title}
-                />
+                    <NavigationHeaderLargeTitle
+                        heightOffset={heightOffset}
+                        hasSearch={hasSearch}
+                        subtitle={subtitle}
+                        theme={theme}
+                        title={title}
+                        translateY={translateY}
+                    />
                 }
                 {hasSearch &&
                     <NavigationSearch
                         {...searchProps}
-                        defaultHeight={defaultHeight}
```

Comment:
I added a new prop `topMargin` and calculate the value based on the props that are available at this level (`defaultHeight`, `largeHeight`, `scrollValue`)